jaxley/modules/network.py

```diff
@@ -127,28 +135,108 @@ def __getattr__(self, key: str):
         else:
             raise KeyError(f"Key {key} not recognized.")
 
-    def init_morph(self):
+    def _init_morph_custom_spsolve(self):
         self.branchpoint_group_inds = build_branchpoint_group_inds(
             len(self.par_inds),
             self.child_belongs_to_branchpoint,
-            self.nseg,
-            self.total_nbranches,
+            self.cumsum_nseg[-1],
         )
         self.children_in_level = merge_cells(
             self.cumsum_nbranches,
-            self.cumsum_nbranchpoints,
+            self.cumsum_nbranchpoints_per_cell,
             [cell.children_in_level for cell in self.cells],
             exclude_first=False,
         )
         self.parents_in_level = merge_cells(
             self.cumsum_nbranches,
-            self.cumsum_nbranchpoints,
+            self.cumsum_nbranchpoints_per_cell,
             [cell.parents_in_level for cell in self.cells],
             exclude_first=False,
         )
-        self.initialized_morph = True
+        self.root_inds = self.cumsum_nbranches[:-1]
+
+    def _init_morph_jax_spsolve(self):
+        """Initialize the morphology for networks.
+
+        The reason that this is a bit involved is that Jaxley considers branchpoint
+        nodes to be at the very end of __all__ nodes (i.e. the branchpoints of the
+        first cell are even after the compartments of the second cell. The reason for
+        this is that, otherwise, `cumsum_nseg` becomes tricky).
+
+        To achieve this, we first loop over all compartments and append them, and then
+        loop over all branchpoints and append those. The code for building the indices
+        from the `comp_edges` is identical to `jx.Cell`.
+        """
+        self._internal_node_inds = np.arange(self.cumsum_nseg[-1])
+        self._cumsum_nseg_per_cell = jnp.concatenate(
+            [
+                jnp.asarray([0]),
+                jnp.cumsum(jnp.asarray([cell.cumsum_nseg[-1] for cell in self.cells])),
+            ]
+        )
+
+        self._comp_edges = pd.DataFrame()
+
+        # Add all the internal nodes.
+        for offset, cell in zip(self._cumsum_nseg_per_cell, self.cells):
```

Comment:
It would make the ordering in `comp_edges` less intuitive. And I think if we sort afterwards we would have to be very careful to not mess up within-brach-sorting. I would prefer to keep it as it is for now.